keystore/src/connection/platform/wasm/mod.rs

```diff
@@ -99,113 +101,29 @@ impl DatabaseConnection for WasmConnection {
-                    .auto_increment(false)
-                    .add_index(IndexBuilder::new("id".into(), KeyPath::new_single("id")).unique(true)),
-            )
-            .add_object_store(
-                ObjectStoreBuilder::new("e2ei_refresh_token")
-                    .auto_increment(false)
-                    .add_index(IndexBuilder::new("id".into(), KeyPath::new_single("id")).unique(true)),
-            )
-            .add_object_store(
-                ObjectStoreBuilder::new("e2ei_acme_ca")
-                    .auto_increment(false)
-                    .add_index(IndexBuilder::new("id".into(), KeyPath::new_single("id")).unique(true)),
-            )
-            .add_object_store(
-                ObjectStoreBuilder::new("e2ei_intermediate_certs")
-                    .auto_increment(false)
-                    .add_index(
-                        IndexBuilder::new("ski_aki_pair".into(), KeyPath::new_single("ski_aki_pair")).unique(true),
-                    ),
-            )
-            .add_object_store(ObjectStoreBuilder::new("e2ei_crls").auto_increment(false).add_index(
-                IndexBuilder::new("distribution_point".into(), KeyPath::new_single("distribution_point")).unique(true),
-            ))
-            .add_object_store(
-                ObjectStoreBuilder::new("proteus_prekeys")
-                    .auto_increment(false)
-                    .add_index(IndexBuilder::new("id".into(), KeyPath::new_single("id")).unique(true)),
-            )
-            .add_object_store(
-                ObjectStoreBuilder::new("proteus_identities")
-                    .auto_increment(false)
-                    .add_index(IndexBuilder::new("pk".into(), KeyPath::new_single("pk")).unique(true)),
-            )
-            .add_object_store(
-                ObjectStoreBuilder::new("proteus_sessions")
-                    .auto_increment(false)
-                    .add_index(IndexBuilder::new("id".into(), KeyPath::new_single("id")).unique(true)),
-            );
-
-        #[cfg(feature = "idb-regression-test")]
-        let idb_builder =
-            idb_builder.add_object_store(ObjectStoreBuilder::new("regression_check").auto_increment(false));
-
-        let idb = idb_builder.build().await?;
+        let factory = Factory::new()?;
+
+        let open_existing = factory.open(&name, None)?;
+        let existing_db = open_existing.await?;
+        let mut migrated_version = existing_db.version()?;
+
+        let idb = if migrated_version == version {
+            // Migration is not needed, just return existing db
+            existing_db
+        } else {
+            // Migration is needed
+            existing_db.close();
+
+            while migrated_version < version {
+                migrated_version = migrate(migrated_version, version, &name, key).await?;
+            }
+
+            let open_request = factory.open(&name, Some(version))?;
+            open_request.await?
+        };
```

Comment:
I think this block that iteratively calls `migrate` should be better placed in the `migration` module (also: consider renaming the module to `migrations` -- if it's likely there will be more than one :wink: ).

So here we wouldn't need to be concerned about how migrations are performed, we just delegate that the migration module and continue with our work:
```rust

        let db = factory.open(&name, None)?.await?;
        let db = migrate(db); // takes ownership of `db`
        ...
```
In this case the pub `migrate` function would be the "driver" of all migrations and the looping is just a migration implementation detail.